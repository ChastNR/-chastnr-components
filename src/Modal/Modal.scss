@-webkit-keyframes show-overlay-animation {
  0% {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@keyframes show-overlay-animation {
  0% {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@-webkit-keyframes show-popup-fade-animation {
  0% {
    opacity: 0;
    -webkit-transform: translateY(120px);
    transform: translateY(120px);
  }

  to {
    opacity: 1;
    -webkit-transform: translateY(0);
    transform: translateY(0);
  }
}

@keyframes show-popup-fade-animation {
  0% {
    opacity: 0;
    -webkit-transform: translateY(120px);
    transform: translateY(120px);
  }

  to {
    opacity: 1;
    -webkit-transform: translateY(0);
    transform: translateY(0);
  }
}

.ovrly {
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  height: 100%;
  left: 0;
  position: fixed;
  right: 0;
  top: 0;
  width: 100%;
  z-index: 9999;

  &:before {
    animation: show-overlay-animation 0.5s cubic-bezier(0.54, 1.12, 1, 1) forwards;
    background: rgba(51, 51, 51, 0.24);
    content: '';
    display: block;
    height: 100%;
    opacity: 0;
    width: 100%;
  }

  .ovrly__popup {
    bottom: 0;
    left: 0;
    outline: none;
    overflow-x: auto;
    overflow-y: scroll;
    position: absolute;
    right: 0;
    top: 0;
  }
}

.mdl {
  box-sizing: border-box;
  display: flex;
  flex-grow: 1;
  flex-shrink: 0;
  margin: auto;
  min-height: 100%;
  padding: 40px 44px;
  pointer-events: none;
  width: 100%;

  .mdl__cntr {
    align-self: center;
    background-color: #fff;
    border-radius: 8px;
    box-shadow: 0 18px 30px rgb(51 51 51 / 52%);
    box-sizing: border-box;
    color: rgba(0, 0, 0, 0.8);
    margin-left: auto;
    margin-right: auto;
    max-width: 600px;
    min-height: 40px;
    padding: 40px 0;
    pointer-events: all;
    position: relative;
    width: 100%;
    animation: show-popup-fade-animation 0.3s ease-out;

    & > button {
      appearance: none;
      background: none;
      border: none;
      color: rgba(0, 0, 0, 0.4);
      cursor: pointer;
      outline: none;
      padding: 0;
      pointer-events: all;
      position: absolute;
      right: 8px;
      top: 8px;
      transition: color 0.2s ease;
    }

    & > div {
      display: flex;
      flex-direction: column;
      min-width: 100%;
      position: relative;
    }
  }
}
