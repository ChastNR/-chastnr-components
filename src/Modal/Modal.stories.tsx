import { useState } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { ComponentMeta } from '@storybook/react';
import { Modal } from './Modal';

const componentMeta: ComponentMeta<typeof Modal> = {
  title: 'Modal',
  component: Modal,
  parameters: {
    layout: 'fullscreen',
  },
};

export const DefaultModal = () => {
  const [isVisible, setVisible] = useState(false);

  return (
    <div>
      <button onClick={() => setVisible(true)}>Show</button>
      {isVisible && <Modal onClose={() => setVisible(false)}>Test Modal</Modal>}
    </div>
  );
};

export default componentMeta;
