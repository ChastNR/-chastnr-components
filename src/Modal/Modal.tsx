import { PropsWithChildren, useRef } from 'react';
import classNames from 'clsx';
import { Close } from '../base';
import { useEscapeKeyDown } from '../hooks';

import './Modal.scss';

interface IModalProps
  extends PropsWithChildren<{
    className?: string;
    onClose: (e?: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  }> {}

export const Modal: React.FC<IModalProps> = ({ children, className, onClose }) => {
  const modalRef = useRef<HTMLDivElement>(null);

  useEscapeKeyDown(() => onClose());

  const handleOverlayClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    if (modalRef.current?.contains(e.target as Node)) {
      return;
    }

    onClose(e);
  };

  const handleButtonClick = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.stopPropagation();
    onClose(e);
  };

  const modalClassNames = classNames('mdl', className);

  return (
    <div className="ovrly">
      <div className="ovrly__popup" onClick={handleOverlayClick} role="presentation">
        <div className={modalClassNames}>
          <div className="mdl__cntr" ref={modalRef}>
            <button onClick={handleButtonClick} type="button">
              <Close />
            </button>
            <div>{children}</div>
          </div>
        </div>
      </div>
    </div>
  );
};
