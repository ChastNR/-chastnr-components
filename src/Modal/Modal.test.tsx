import { fireEvent, render } from '@testing-library/react';
import { Modal } from './Modal';

describe('Modal', () => {
  const mockOnClose = jest.fn();

  it('should render properly', () => {
    const modal = render(
      <Modal className="testClassName" onClose={mockOnClose}>
        Test content
      </Modal>
    );

    expect(modal.container.firstChild).toMatchSnapshot();
  });

  it('should not close modal when modal was clicked', () => {
    jest.mock('../hooks', () => ({
      useEscapeKeyDown: jest.fn(),
    }));

    const modal = render(
      <Modal className="testClassName" onClose={mockOnClose}>
        Test content
      </Modal>
    );

    const overlayElement = modal.container.getElementsByClassName('mdl')[0];

    fireEvent.click(overlayElement);

    // useEffect dev problem
    expect(mockOnClose).toBeCalledTimes(1);
  });

  it('should close modal when overlay was clicked', () => {
    const modal = render(
      <Modal className="testClassName" onClose={mockOnClose}>
        Test content
      </Modal>
    );

    const overlayElement = modal.container.getElementsByClassName('ovrly__popup')[0];

    fireEvent.click(overlayElement);

    // useEffect dev problem
    expect(mockOnClose).toBeCalledTimes(2);
  });

  it('should close modal when close button was clicked', () => {
    const modal = render(
      <Modal className="testClassName" onClose={mockOnClose}>
        Test content
      </Modal>
    );

    const buttonElement = modal.container.getElementsByTagName('button')[0];

    fireEvent.click(buttonElement);

    expect(mockOnClose).toBeCalled();
  });

  it('should close modal when ESC was pressed', () => {
    const modal = render(
      <Modal className="testClassName" onClose={mockOnClose}>
        Test content
      </Modal>
    );

    fireEvent.keyDown(document, { key: 'Escape' });

    expect(mockOnClose).toBeCalled();
  });
});
