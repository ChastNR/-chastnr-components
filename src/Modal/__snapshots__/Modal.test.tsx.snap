// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render properly 1`] = `
<div
  class="ovrly"
>
  <div
    class="ovrly__popup"
    role="presentation"
  >
    <div
      class="mdl testClassName"
    >
      <div
        class="mdl__cntr"
      >
        <button
          type="button"
        >
          <span
            class="close"
          >
            <span>
              content
            </span>
          </span>
        </button>
        <div>
          Test content
        </div>
      </div>
    </div>
  </div>
</div>
`;
